### PR TITLE
Flash layout extensions

### PR DESCRIPTION
**Pull request by mbolivar**
This builds upon work done by @nvlsianpu  and @nashif in https://github.com/zephyrproject-rtos/zephyr/pull/1562 by adding equivalent DTS properties and soc.h defines for STM32F4, STM32L4 and NXP K6X.

In addition, the last two patches:

1. select FLASH_PAGE_LAYOUT on nRF targets (2176afc)
2. extend the page layout API to add an iterator over pages (e9d0236)

These are useful to extend mcuboot to obtain the target configuration from Zephyr directly.

With link time garbage collection, 1. should add negligible size to the nRF flash driver when the page layout API is not used (flash storage for flash_nrf5_pages_layout()'s code and the pointer to it in the API struct, RAM for one struct flash_pages_layout, and a few extra instructions in nrf5_flash_init()). Hopefully that's acceptable, as it makes it much simpler to allow applications to use the API whenever it's available.

The iterator in 2. is useful for efficiently converting Zephyr's flash layout into what mcuboot expects at bootloader startup time.

### CODE REVIEW

File: drivers/flash/flash_page_layout.c [hunk 1]
```diff
@@ -80,3 +80,31 @@ size_t flash_get_page_count(struct device *dev)
 
 	return count;
 }
+
+void flash_page_foreach(struct device *dev, flash_page_cb cb, void *data)
```
Comment (nashif) [inline drivers/flash/flash_page_layout.c]: need some code in Zephyr that uses this functions, preferably a testcase...

**Review by nashif — CHANGES_REQUESTED** [review]

**Review by nvlsianpu — CHANGES_REQUESTED** [review]

File: arch/arm/soc/st_stm32/stm32f4/soc.h [hunk 2]
```diff
@@ -48,6 +48,12 @@
 #include <stm32f4xx_ll_rng.h>
 #endif
 
+/* For IMG_MANAGER */
+#if defined(CONFIG_SOC_FLASH_STM32)
+#define FLASH_DRIVER_NAME	CONFIG_SOC_FLASH_STM32_DEV_NAME
+#define FLASH_ALIGN		FLASH_WRITE_BLOCK_SIZE
```
Comment (galak) [inline arch/arm/soc/st_stm32/stm32f4/soc.h]: where is FLASH_ALIGN used?

File: drivers/flash/flash_page_layout.c (hunk 1 above)
Comment (galak) [reply drivers/flash/flash_page_layout.c]: I would drop this from the PR until you have something that uses it.

Comment (mbolivar) [reply drivers/flash/flash_page_layout.c]: I'm using it out of tree, in mcuboot. I'll add a test case.

File: arch/arm/soc/st_stm32/stm32f4/soc.h (hunk 2 above)
Comment (mbolivar) [reply arch/arm/soc/st_stm32/stm32f4/soc.h]: In mcuboot.

File: drivers/flash/Kconfig [hunk 3]
```diff
@@ -127,6 +127,7 @@ config SOC_FLASH_NRF5
 	bool "Nordic Semiconductor nRF5X flash driver"
 	depends on FLASH && SOC_FAMILY_NRF5
 	default n
+	select FLASH_PAGE_LAYOUT
```
[the remark this answers is not in the document]
Comment (mbolivar) [reply drivers/flash/Kconfig]: Okay, that works too.

[the remark this answers is not in the document]
Comment (mbolivar) [reply drivers/flash/Kconfig]: I've dropped this patch.

File: arch/arm/soc/st_stm32/stm32f4/soc.h (hunk 2 above)
Comment (mbolivar) [reply arch/arm/soc/st_stm32/stm32f4/soc.h]: I added a patch which removes FLASH_ALIGN in Zephyr and just uses FLASH_WRITE_BLOCK_SIZE instead. This involved going back and taking it out of soc.h as was done by @nashif and @nvlsianpu in https://github.com/zephyrproject-rtos/zephyr/pull/1562. Hope that's OK.

**Comment by mbolivar** [comment]
Hi, I believe I've addressed open comments.

**Comment by galak** [comment]
why aren't we getting the IMG_MANAGER stuff from dt instead.

**Comment by mbolivar** [comment]
> why aren't we getting the IMG_MANAGER stuff from dt instead.

Sorry, I missed this comment.

Are you saying you want the defines, like the one in https://github.com/zephyrproject-rtos/zephyr/pull/1594/commits/8342256a1ecd69bbc328b343d399ca609067169a, to come from device tree instead?

I.e. we should remove CONFIG_SOC_FLASH_some_soc_NAME from Kconfig and put it in DT instead? Perhaps as a compatible string? Sorry for the confusion.

**Comment by mbolivar** [comment]
> I.e. we should remove CONFIG_SOC_FLASH_some_soc_NAME from Kconfig and put it in DT instead? Perhaps as a compatible string? Sorry for the confusion.

@galak this question is for you

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
I guess we can take it as is right now and see how to make this better work with DT in a followup PR.

**Pull request merged**